skops/io/_numpy.py

```diff
@@ -50,39 +49,57 @@ def ndarray_get_state(obj: Any, save_context: SaveContext) -> dict[str, Any]:
     return res
 
 
-def ndarray_get_instance(state, load_context: LoadContext):
-    # Dealing with a regular numpy array, where dtype != object
-    if state["type"] == "numpy":
-        val = np.load(
-            io.BytesIO(load_context.src.read(state["file"])), allow_pickle=False
-        )
-        # Coerce type, because it may not be conserved by np.save/load. E.g. a
-        # scalar will be loaded as a 0-dim array.
-        if state["__class__"] != "ndarray":
-            cls = _import_obj(state["__module__"], state["__class__"])
-            val = cls(val)
-        return val
-
-    # We explicitly set the dtype to "O" since we only save object arrays in
-    # json.
-    shape = get_instance(state["shape"], load_context)
-    tmp = [get_instance(s, load_context) for s in state["content"]]
-    # TODO: this is a hack to get the correct shape of the array. We should
-    # find _a better way_ to do this.
-    if len(shape) == 1:
-        val = np.ndarray(shape=len(tmp), dtype="O")
-        for i, v in enumerate(tmp):
-            val[i] = v
-    else:
-        val = np.array(tmp, dtype="O")
-    return val
+class NdArrayNode(Node):
+    def __init__(self, state, load_context: LoadContext, trusted=False):
+        super().__init__(state, load_context, trusted)
+        self.type = state["type"]
+        self.trusted = self._get_trusted(trusted, ["numpy.ndarray"])
+        if self.type == "numpy":
+            self.content = io.BytesIO(
+                load_context.src.read(state["file"])
+            )  # type: ignore
+            self.children = {}
+        elif self.type == "json":
+            self.shape = get_tree(state["shape"], load_context)
+            self.content = [
+                get_tree(o, load_context) for o in state["content"]  # type: ignore
+            ]
+            self.children = {"shape": Node, "content": list}
+        else:
+            raise ValueError(f"Unknown type {self.type}.")
```

Comment:
this only should happen if the given file is malformed, which we're not testing for now.